giraph-core/src/main/java/org/apache/giraph/writable/kryo/GiraphClassResolver.java

```diff
@@ -0,0 +1,333 @@
+import com.esotericsoftware.kryo.util.ObjectMap;
+import org.apache.giraph.zk.ZooKeeperExt;
+import org.apache.zookeeper.CreateMode;
+import org.apache.zookeeper.KeeperException;
+import org.apache.zookeeper.ZooDefs;
+import java.util.HashMap;
+import java.util.Map;
+import java.util.List;
+
+import static com.esotericsoftware.kryo.util.Util.getWrapperClass;
+
+/**
+ * This class resolver assigns unique classIds for every class that was not
+ * explicitly registered. It uses zookeeper for consistent mapping across all
+ * nodes.
+ */
+public class GiraphClassResolver extends DefaultClassResolver {
+  /** Base ID to start for class name assignments.
+   * This number has to be high enough to not conflict with
+   * explicity registered class IDs.
+   * */
+  private static final int BASE_CLASS_ID = 1000;
+
+  /** Class name to ID cache */
+  private static Map<String, Integer> CLASS_NAME_TO_ID = new HashMap();
+  /** ID to class name cache */
+  private static Map<Integer, String> ID_TO_CLASS_NAME = new HashMap();
+  /** Zookeeper */
+  private static ZooKeeperExt ZK;
+  /** Zookeeper path for automatic class registrations */
+  private static String KRYO_REGISTERED_CLASS_PATH;
+  /** Minimum class ID assigned by zookeeper sequencing */
+  private static int MIN_CLASS_ID = -1;
+
+  /** Memoized class id*/
+  private int memoizedClassId = -1;
+  /** Memoized class registration */
+  private Registration memoizedClassIdValue;
+
+  /**
+   * Sets zookeeper informaton.
+   * @param zookeeperExt ZookeeperExt
+   * @param kryoClassPath Zookeeper directory path where class Name-ID
+   *                      mapping is stored.
+   */
+  public static void setZookeeperInfo(ZooKeeperExt zookeeperExt,
+                                      String kryoClassPath) {
+    ZK = zookeeperExt;
+    KRYO_REGISTERED_CLASS_PATH = kryoClassPath;
+  }
+
+  /**
+   * Creates a new node for the given class name.
+   * Creation mode is persistent sequential, i.e.
+   * ZK will always create a new node . There could be
+   * multiple entries for the same class name but since
+   * the lowest index is used, this is not a problem.
+   * @param className Class name
+   */
+  public static void createClassName(String className) {
+    try {
+      String path = KRYO_REGISTERED_CLASS_PATH + "/" + className;
+      ZK.createExt(path,
+              null,
+              ZooDefs.Ids.OPEN_ACL_UNSAFE,
+              CreateMode.PERSISTENT_SEQUENTIAL,
+              true);
+    } catch (KeeperException e) {
+      throw new IllegalStateException(
+              "Failed to create class " + className, e);
+    } catch (InterruptedException e) {
+      throw new IllegalStateException(
+              "Interrupted while creating " + className, e);
+    }
+  }
+
+  /**
+   * Refreshes class-ID mapping from zookeeper.
+   * Not thread safe.
+   */
+  public static void refreshCache() {
+    try {
+      ZK.createOnceExt(KRYO_REGISTERED_CLASS_PATH,
```

Comment:
I think you can avoid calling this every time you refresh the cache and, instead, call it once during initialization (e.g. in setZookeeperInfo.

Reply:
I have added a boolean flag to avoid this repeated call.